docs/DesignDocs/AttestationAPIsWithFlexibleClaims.md

```diff
@@ -0,0 +1,816 @@
+  support (e.g., `CONFIGID` in SGX2). The `oe_get_attestation_certificate_with_evidence_v2`
+  API does not allow developers to provide an init-time custom claim buffer as input.
+
+- Limited support for run-time custom claims
+
+  The notion of run-time custom claims is equivalent to the custom claims
+  in the OE attestation plugin framework. The `oe_get_attestation_certificate_with_evidence_v2`
+  API does not allow developers to provide custom claims as input.
+  Instead, the API internally takes the input public key buffer as the custom claim by default.
+  The `oe_verify_attestation_certificate_with_evidence_v2` then extracts the custom
+  claim and verifies its value against the public key.
+
+Specification
+--------
+
+This document proposes new APIs,
+`oe_get_attestation_certificate_with_evidence_v3` and
+`oe_parse_attestation_ceritiface_with_evidence`, that
+address the abovementioned limitations.
+Note that the certificate generated by `oe_get_attestation_certificate_with_evidence_v3`
+does not provide backward compatibility to existing verification APIs
+(e.g., `oe_verify_attestation_certificate_with_evidence_v2`
+and `oe_verify_attestation_certificate`) if the run-time custom claims do not include
+only the public key.
+
+- oe_get_attestation_certificate_with_evidence_v3
+  ```c
+  /**
+   * oe_get_attestation_certificate_with_evidence_v3
+   *
+   * Similar to oe_get_attestation_certificate_with_evidence_2, this function
+   * generates a self-signed X.509 certificate with embedded evidence generated by
+   * an attester plugin for the enclave. In addition, it allows an application to optionally
+   * pass in run-time custom claims, init-time custom claims, and attestation results.
+   *
+   * @experimental
+   *
+   * @param[in] format_id The format id of the evidence to be generated.
+   * @param[in] subject_name A string containing an X.509 distinguished
+   * name (DN) for customizing the generated certificate. This name is also used
+   * as the issuer name because this is a self-signed certificate.
+   * See RFC5280 (https://tools.ietf.org/html/rfc5280) for details.
+   * Example value "CN=Open Enclave SDK,O=OESDK TLS,C=US"
+   * @param[in] private_key A private key used to sign this certificate.
+   * @param[in] private_key_size The size of the buffer above.
+   * @param[in] public_key A public key used as the certificate's subject key.
+   * @param[in] public_key_size The size of the buffer above.
+   * @param[in] inittime_custom_claims_buffer An optional flat buffer representing the init-time
+   * custom claims.
+   * @param[in] inittime_custom_claims_buffer_size The size of the buffer above in bytes.
+   * @param[in] runtime_custom_claims_buffer An optional flat buffer representing the run-time
+   * cutom claims.
+   * @param[in] runtime_custom_claims_buffer_size The size of the buffer above in bytes.
+   * @param[in] attestation_results_buffer An optional flat buffer representing the attestation
```

Comment:
Not sure how adding `attestatin_results_buffer` to this `oe_get_attestation_certificate_with_evidence_v3()` API is the best approach to add support of the passport mode. Does this API always include evidence in the output cert? But in passport mode, the attestation results is supposed to be carried instead of the evidence.

Reply:
You're correct, the example of passport mode below does not require evidence to be embedded. The APIs probably need different names.